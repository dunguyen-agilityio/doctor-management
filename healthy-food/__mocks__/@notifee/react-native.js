export default {
  requestPermission: jest.fn(() => Promise.resolve(true)),
  displayNotification: jest.fn(() => Promise.resolve()),
  onForegroundEvent: jest.fn(),
  onBackgroundEvent: jest.fn(),
  createChannel: jest.fn(() => Promise.resolve('mock-channel')),
  getInitialNotification: jest.fn(() => Promise.resolve(null)),
};
