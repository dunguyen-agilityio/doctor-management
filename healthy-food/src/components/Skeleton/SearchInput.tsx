import { StyleSheet } from 'react-native';

import Skeleton from './Skeleton';

const SearchInputSkeleton = () => {
  return <Skeleton height={46} shimmerStyle={styles.container} />;
};

export default SearchInputSkeleton;

const styles = StyleSheet.create({
  container: {
    width: '100%',
  },
});
