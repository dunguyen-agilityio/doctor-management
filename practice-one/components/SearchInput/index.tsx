import { StyleSheet, TextInput, View } from 'react-native';
import React, { useCallback, useRef } from 'react';

import { COLORS } from '@constants';
import { Input } from '@components/common';
import { SearchIcon } from '@constants';

import { useFilterStore } from '@stores/filter';

interface SearchInputProps {
  onFocus?: () => void;
}

const SearchInput = ({ onFocus }: SearchInputProps) => {
  const ref = useRef<TextInput>(null);

  const { query, setFilter } = useFilterStore();

  const handleChangeQuery = useCallback(
    (query: string) => {
      setFilter({ query });
    },
    [setFilter],
  );

  return (
    <View style={styles.container}>
      <Input
        field="search"
        onChangeText={handleChangeQuery}
        value={query}
        placeholder="Search for healthy food"
        ref={ref}
        onFocus={onFocus}
      />

      <View style={styles.iconSearch}>
        <SearchIcon />
      </View>
    </View>
  );
};

export default SearchInput;

const styles = StyleSheet.create({
  container: {
    position: 'relative',
    backgroundColor: COLORS.LIGHT_PURPLE,
    borderRadius: 15,
    zIndex: 1,
    marginTop: 14,
    marginHorizontal: 16,
  },
  text: {
    fontSize: 13,
    fontWeight: '400',
    color: COLORS.WARNING,
    paddingVertical: 14,
    paddingHorizontal: 60,
  },
  iconSearch: {
    position: 'absolute',
    left: 21,
    top: '50%',
    transform: [{ translateY: -10 }],
  },
});
