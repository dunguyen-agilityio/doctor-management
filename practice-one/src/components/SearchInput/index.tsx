import { forwardRef, useImperativeHandle, useRef } from 'react';
import { StyleSheet, TextInput, TextInputProps, View } from 'react-native';

import MaterialIcons from '@expo/vector-icons/MaterialIcons';
import { Image } from 'expo-image';

import { APP_ICONS, COLOR } from '@/constants';

import { debounce } from '@/utils/debounce';

import Button from '../Button';

interface SearchInputProps extends TextInputProps {
  onSearch?: (value: string) => void;
  query?: string;
}

const SearchInput = (
  { onSearch, onPress, query, ...otherProps }: SearchInputProps,
  ref: React.ForwardedRef<Pick<TextInput, 'focus' | 'clear'>>,
) => {
  const inputRef = useRef<TextInput>(null);

  useImperativeHandle(
    ref,
    () => ({
      focus: () => inputRef.current?.focus(),
      clear: () => {
        return inputRef.current?.clear();
      },
    }),
    [],
  );

  const debouncedChangeText = debounce((value: string) => {
    onSearch?.(value);
  }, 500);

  const handleClear = () => {
    onSearch?.('');
    inputRef.current?.clear();
  };

  return (
    <View style={styles.container} testID="search-input">
      <Image source={APP_ICONS.SEARCH} style={styles.iconSearch} />
      <TextInput
        {...otherProps}
        ref={inputRef}
        placeholder="Search for healthy food"
        placeholderTextColor={COLOR.SECONDARY}
        style={styles.input}
        defaultValue={query}
        onChangeText={debouncedChangeText}
      />
      {query && (
        <Button
          variant="icon"
          width={12}
          onPress={handleClear}
          style={styles.clearButton}
          backgroundColor="transparent"
          testID="clear-button"
        >
          <MaterialIcons name="clear" size={12} color={COLOR.RED} />
        </Button>
      )}
    </View>
  );
};

export default forwardRef(SearchInput);

const styles = StyleSheet.create({
  container: {
    position: 'relative',
    backgroundColor: COLOR.LIGHT_PURPLE,
    borderRadius: 15,
    marginTop: 14,
    marginHorizontal: 16,
    height: 46,
  },
  input: {
    fontSize: 13,
    fontWeight: '400',
    paddingVertical: 14,
    paddingHorizontal: 60,
    color: COLOR.BLACK,
  },
  iconSearch: {
    position: 'absolute',
    left: 21,
    top: '50%',
    transform: [{ translateY: -10 }],
    width: 24,
    height: 24,
  },
  clearButton: {
    position: 'absolute',
    right: 16,
    top: '50%',
    transform: [{ translateY: '-50%' }],
  },
});
