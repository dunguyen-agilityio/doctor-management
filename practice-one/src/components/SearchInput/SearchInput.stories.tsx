import type { Meta, StoryObj } from '@storybook/react';

import { View } from 'react-native';

import SearchInput from './index';

const meta = {
  title: 'SearchInput',
  component: SearchInput,
  argTypes: { onChangeText: { action: 'type' } },
  decorators: [
    (Story) => (
      <View style={{ flex: 1 }}>
        <Story />
      </View>
    ),
  ],
} satisfies Meta<typeof SearchInput>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const WithQuery: Story = { args: { query: 'abc' } };
