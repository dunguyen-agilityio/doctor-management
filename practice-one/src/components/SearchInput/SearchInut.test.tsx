import React, { createRef } from 'react';
import { TextInput } from 'react-native';

import { fireEvent, render, waitFor } from '@testing-library/react-native';

import SearchInput from '../SearchInput';

describe('SearchInput Component', () => {
  it('renders correctly with placeholder', () => {
    const { getByPlaceholderText } = render(<SearchInput />);
    const input = getByPlaceholderText('Search for healthy food');
    expect(input).toBeTruthy();
  });

  it('calls onChangeText when typing', async () => {
    const mockOnChangeText = jest.fn();
    const { getByPlaceholderText } = render(
      <SearchInput onChangeText={mockOnChangeText} />,
    );

    const input = getByPlaceholderText('Search for healthy food');
    fireEvent.changeText(input, 'apple');

    await waitFor(() => expect(mockOnChangeText).toHaveBeenCalledWith('apple'));
  });

  it('focuses when tapped anywhere inside the container', () => {
    const inputRef = createRef<TextInput>();
    const { getByTestId } = render(<SearchInput ref={inputRef} />);

    fireEvent.press(getByTestId('search-input'));

    expect(inputRef.current?.focus).toBeDefined();
  });

  it('matches snapshot', () => {
    const tree = render(<SearchInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
