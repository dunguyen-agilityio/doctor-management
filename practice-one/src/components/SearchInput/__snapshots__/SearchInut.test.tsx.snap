// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Component matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFF3F1",
      "borderRadius": 15,
      "marginHorizontal": 16,
      "marginTop": 14,
      "position": "relative",
      "zIndex": 1,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Search for healthy food"
    placeholderTextColor="#FF8473"
    style={
      {
        "color": "#000000",
        "fontSize": 13,
        "fontWeight": "400",
        "paddingHorizontal": 60,
        "paddingVertical": 14,
      }
    }
    testID="search-input"
  />
  <View
    style={
      {
        "left": 21,
        "position": "absolute",
        "top": "50%",
        "transform": [
          {
            "translateY": -10,
          },
        ],
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../src/assets/icons/iconsearch.png",
        }
      }
    />
  </View>
</View>
`;
